### PR TITLE
2022 Week 44 release

### PR DESCRIPTION
**Pull request by mr-c**

### CODE REVIEW

File: src/topics/command-line-tool.md [hunk 1]
```diff
@@ -63,3 +63,23 @@ and in the [Outputs](../topics/outputs.md) sections.
 %
 % - Spaces in commands https://github.com/common-workflow-language/user_guide/issues/39
 % - Arguments (tell the reader the different use cases for arguments and inputs, tell them there is a section about inputs)
+
+
+## Network Access
+This indicates whether a process requires outgoing IPv4/IPv6 network access.
+If a command-line tool is written manually in CWL v1.1+, there is a need to 
+specify when network access is required.
```
Comment (mr-c) [inline src/topics/command-line-tool.md:71]:
```suggestion
Starting with CWL v1.1, you must document the requirement for network access if it is required.
```

File: src/topics/command-line-tool.md [hunk 2]
```diff
@@ -63,3 +63,23 @@ and in the [Outputs](../topics/outputs.md) sections.
...
+class: CommandLineTool
+
+requirements:
+  NetworkAccess:
+     networkAccess: true
+```
+
+```{note}
+CWL v1.0 command-line tools that are upgraded to v1.1 
+or v1.2 get Network Access automatically.
```
Comment (mr-c) [inline src/topics/command-line-tool.md:84]:
```suggestion
or v1.2 get `networkAccess: true` set automatically.
```

File: src/topics/command-line-tool.md [hunk 3]
```diff
@@ -63,3 +63,23 @@ and in the [Outputs](../topics/outputs.md) sections.
 %
 % - Spaces in commands https://github.com/common-workflow-language/user_guide/issues/39
 % - Arguments (tell the reader the different use cases for arguments and inputs, tell them there is a section about inputs)
+
+
+## Network Access
+This indicates whether a process requires outgoing IPv4/IPv6 network access.
```
Comment (mr-c) [inline src/topics/command-line-tool.md:69]: Hmm, the placement of this new section seems awkward. None of the other CommandLineTool specific requirements are documented in this section.

I think a new section about network access is fine. We should also explain why network access using steps should be avoided: hurts portability, increases the fragility of the workflow (external service can go away, or change format of responses). We should also present alternatives like putting the URL for an input File in the input object's `location` field instead of a `curl` or `wget` using `CommandLineTool`; and in general separating the automated batch steps from data acquisition and data publishing steps.

File: src/introduction/prerequisites.md [hunk 4]
```diff
@@ -26,13 +26,13 @@ of the following options for your operating system:
 - Windows
 
 ```{note}
-If you are using Windows, you will have to install the Windows Subsystem for Linux 2.
+If you are using Windows, you will have to install the [Windows Subsystem for Linux 2](https://learn.microsoft.com/en-us/windows/wsl/install) (WSL2).
 Visit the `cwltool` [documentation](https://github.com/common-workflow-language/cwltool/blob/main/README.rst#ms-windows-users)
 for details on installing WSL2.
```
Comment (mr-c) [inline src/introduction/prerequisites.md:31]: Best not to link directly to the WSL2 install docs, as users will click that first and not follow all of the linked cwltool MS Windows installation directions

```suggestion
If you are using Windows, you will have to install the Windows Subsystem for Linux 2 as documented in the
[`cwltool` documentation for Microsoft Windows users](https://github.com/common-workflow-language/cwltool/blob/main/README.rst#ms-windows-users).
```

File: src/introduction/prerequisites.md [hunk 5]
```diff
@@ -52,29 +52,29 @@ environment:
...
-$ python -m venv venv
+$ python3 -m venv venv
 $ source venv/bin/activate
 $ (venv) pip install -U pip setuptools wheel
 $ (venv) pip install cwltool
 ```
 
 ```{note}
-You can find the `cwl-runner` source code [here](https://github.com/common-workflow-language/cwltool/tree/main/cwlref-runner).
+You can find the `cwl-runner` [source code](https://github.com/common-workflow-language/cwltool/tree/main/cwlref-runner) in the `cwltool` repository.
```
Comment (mr-c) [inline src/introduction/prerequisites.md:62]: This entire line should be deleted

**Review by mr-c — COMMENTED** [review]
My initial review.

It would be nice if more of this was caught during the original PRs, @tetron 

Also, the commit messages are very uninformative. Either PRs should be squashed and the commit message updated, or contributors should be asked to clean up their commits.

File: src/introduction/prerequisites.md (hunk 5 above)
Comment (swzCuroverse) [reply src/introduction/prerequisites.md:62]: There was direct ask from Bruno to add this in.  And then it is better to indicated what the link is instead of saying "here".

File: src/topics/command-line-tool.md (hunk 3 above)
Comment (swzCuroverse) [reply src/topics/command-line-tool.md:69]: Could we add this now and then make a ticket to add this information.  The ticket just said to add this specifically.   https://github.com/common-workflow-language/user_guide/issues/249

File: src/faq.md [hunk 6]
```diff
@@ -8,7 +8,7 @@
 :backlinks: "top"
 ```
 
-## Non "`File`" types using `evalFrom`
+## Non "`File`" Types Using `evalFrom`
```
[the remark this answers is not in the document]
Comment (swzCuroverse) [reply src/faq.md:11]: Non "`File`" Types Using `evalFrom` --> this is not a question

File: src/faq.md [hunk 7]
```diff
@@ -191,7 +246,7 @@ inputs:
 <a name="enuminputs"></a>
 ## Enum Inputs ⚜️
 
-For command-line flags that require a specific input as the argument, an enum type can be declared in CWL. **Specifying null here is known as long form style. It does the same thing as the question mark on the other inputs.**
+For command line flags that require a specific input as the argument an enum type can be declared in CWL. **Specifying null here is known as long form style. It does the same thing as the question mark on the other inputs.**
```
[the remark this answers is not in the document]
Comment (swzCuroverse) [reply src/faq.md:249]: I believe we aren't consistent throughout the whole document.  It would be good to say which in then apply it to everyone equally.  Currently command-line is used throughout the document https://github.com/common-workflow-language/user_guide/search?q=command-line.

File: src/faq.md [hunk 8]
```diff
@@ -79,7 +79,62 @@ outputs:
...
+
+  # other inputs go here
+
+baseCommand: sh
+
+outputs: []
+```
+
+```{note}
+In CWL, everything must be directly stated.
```
[the remark this answers is not in the document]
Comment (swzCuroverse) [reply src/faq.md:134]: I don't understand what you want to add here.

File: src/faq.md [hunk 9]
```diff
@@ -79,7 +79,62 @@ outputs:
       outputEval: ${self[0].basename=inputs.otu_table_name; return self;}
 ```
 
-## Setting `self`-based input bindings for optional inputs
+## Referencing a Local Script
```
[the remark this answers is not in the document]
Comment (swzCuroverse) [reply src/faq.md:82]: Why is this a question?  I think you are saying all FAQs have to be questions.  I think that is a little pedantic. and currently not the case for most of the existing FAQs - so do you want to change all of them?  Additionally you reviewed this with Bruno and previously didn't have an issue- so we shouldn't blame others for doing the same. http://www.commonwl.org/user_guide/faq.html

File: src/faq.md [hunk 10]
```diff
@@ -38,12 +38,12 @@ outputs:
        outputEval: $(self[0].contents)
 ```
 
-## Rename an input file
+## Rename an Input File
```
[the remark this answers is not in the document]
Comment (swzCuroverse) [reply src/faq.md:41]: Why is this a question? I think you are saying all FAQs have to be questions. I think that is a little pedantic. and currently not the case for most of the existing FAQs, none of them are questions.  Some of phrases and some perhaps are sentences.  See here:  http://www.commonwl.org/user_guide/faq.html

File: src/faq.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (swzCuroverse) [reply src/faq.md:11]: I wrote below - if you want FAQ titles to be questions then you have to fix all the existing titles and add questions.   For clarification, current these are the titles that you/Bruno have approved for the FAQs -- not that none of them have ? or sentence capitalization 

[FAQ](https://www.commonwl.org/user_guide/faq.html)

    [4.1. Non “File” types using evalFrom](https://www.commonwl.org/user_guide/faq.html#non-file-types-using-evalfrom)
    [4.2. Rename an input file](https://www.commonwl.org/user_guide/faq.html#rename-an-input-file)
    [4.3. Rename an output file](https://www.commonwl.org/user_guide/faq.html#rename-an-output-file)
    [4.4. Setting self-based input bindings for optional inputs](https://www.commonwl.org/user_guide/faq.html#setting-self-based-input-bindings-for-optional-inputs)
    [4.5. Model a “one-or-the-other” parameter](https://www.commonwl.org/user_guide/faq.html#model-a-one-or-the-other-parameter)
    [4.6. Connect a solo value to an input that expects an array of that type](https://www.commonwl.org/user_guide/faq.html#connect-a-solo-value-to-an-input-that-expects-an-array-of-that-type)
    [4.7. Optional Inputs 💯](https://www.commonwl.org/user_guide/faq.html#optional-inputs)
    [4.8. Enum Inputs ⚜️](https://www.commonwl.org/user_guide/faq.html#enum-inputs)
    [4.9. Record Inputs 📀](https://www.commonwl.org/user_guide/faq.html#record-inputs)
    [4.10. Setting Mutually Exclusive Parameters](https://www.commonwl.org/user_guide/faq.html#setting-mutually-exclusive-parameters)
    [4.11. Setting Booleans](https://www.commonwl.org/user_guide/faq.html#setting-booleans)
    [4.12. Concatenating Strings in Inputs](https://www.commonwl.org/user_guide/faq.html#concatenating-strings-in-inputs)
    [4.13. cwltool errors due to filenames with space characters inside](https://www.commonwl.org/user_guide/faq.html#cwltool-errors-due-to-filenames-with-space-characters-inside)
    [4.14. CWL Parameter Reference error due to hyphen in input identifier](https://www.commonwl.org/user_guide/faq.html#cwl-parameter-reference-error-due-to-hyphen-in-input-identifier)
    [4.15. Use CWL and cwltool with Singularity](https://www.commonwl.org/user_guide/faq.html#use-cwl-and-cwltool-with-singularity)
    [4.16. Debug JavaScript expressions](https://www.commonwl.org/user_guide/faq.html#debug-javascript-expressions)

File: src/faq.md [hunk 11]
```diff
@@ -79,7 +79,62 @@ outputs:
...
+
+```bash
+export PATH=$PATH:/appropriate/directory
+```
+
+Now you can use `baseCommand: scriptname.sh` to run the script directly.
+
+```cwl
+#!/bin/bash
+cwlVersion: v1.0
```
[the remark this answers is not in the document]
Comment (swzCuroverse) [reply src/faq.md:107]: Again, if we except people to follow a particular formatting, then it should be consistent in the User Guide so people know what to do.  I recommend for things like this - you make a list, change them in the user guide and then make sure people can find the list,  perhaps in the contributions.md.  Also, not sure how @tetron or I would have know about what you wanted since it is inconsistent in the User Guide that you approved.

File: src/faq.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (swzCuroverse) [reply src/faq.md:134]: I believe this was taken from rephrasing this comment of yours -- 

"In CWL we have to be explicit about everything, so either the script needs to be

    available on the system PATH (for cwltool doing local execution, no --preserve-environment needed)

or

    part of the Docker format software container (either a fixed path or again on the system PATH defined in the container)

or

    it needs to be an input to the CommandLineTool description itself (as shown in my previous comment)

InitialWorkDirRequirement could be used to arrange the input data alongside one or more scripts (themselves also inputs) if they expect to be in the same folder; is that what you're asking for?"

File: src/faq.md [hunk 12]
```diff
@@ -265,7 +320,7 @@ For command-line flags that are either **mutually exclusive** or **dependent**,
 ```
 ## Setting Mutually Exclusive Parameters
```
[the remark this answers is not in the document]
Comment (swzCuroverse) [reply src/faq.md:321]: Again - this is not how any of the other FAQ titles are currently formatting.  I personally find a longer question harder to parse and doesn't give me any more information.  If you want to add additional content sure, but I don't think we need to make them questions - it is sort of implied by the FAQ title.

[the remark this answers is not in the document]
Comment (swzCuroverse) [reply src/faq.md:321]: We could actually think about removing the FAQ section entirely and trying to weave these into the whole User Guide.  A current school of thought is that FAQs aren't the best idea because they break the narrative and flow and you don't get the information in context.  Something to think about for the future.

File: src/introduction/basic-concepts.md [hunk 13]
```diff
@@ -149,32 +150,32 @@ runners.
...
 > The FAIR principles have laid a foundation for sharing and publishing
-> digital assets and, in particular, data. The FAIR principles emphasize
+> digital assets, and in particular, data. The FAIR principles emphasize
 > machine accessibility and that all digital assets should be Findable,
 > Accessible, Interoperable, and Reusable. Workflows encode the methods
 > by which the scientific process is conducted and via which data are
-> created. It is thus important that workflows both support the creation
-> of FAIR data and themselves adhere to the FAIR principles.
+> created. It is thus important that workflows support the creation
+> of FAIR data and adhere to the FAIR principles.
```
[the remark this answers is not in the document]
Comment (swzCuroverse) [reply src/introduction/basic-concepts.md:163]: " we shouldn't be editing it unless that is really needed"  -> assume you mean we shouldn't be including it/quoting it?
I am on the fence.  If it is a good phrasing from a canonical source, probably better quoting than rephrasing it ourselves.

File: src/faq.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (mr-c) [reply src/faq.md:107]: This was an instruction that Bruno followed, but we didn't document elsewhere:

https://github.com/common-workflow-language/user_guide/issues/222#issuecomment-1172865593

Issue for follow-up: https://github.com/common-workflow-language/user_guide/issues/349

File: src/faq.md [hunk 14]
```diff
@@ -79,7 +79,62 @@ outputs:
...
+#!/bin/bash
+```
+
+After that, make the script executable with the command `chmod +x scriptname.sh`
+
+Finally, modify your `PATH` to add the directory where your script is located.
+(It is good practice to use `$HOME/bin` for storing your own scripts).
+
+```bash
+export PATH=$PATH:/appropriate/directory
```
[the remark this answers is not in the document]
Comment (mr-c) [reply src/faq.md:100]: > > Also, the commit messages are very uninformative. Either PRs should be squashed and the commit message updated, or contributors should be asked to clean up their commits.
> 
> Did we give examples in the contributions guide? If not - perhaps we should since I didn't know we had strict rules about commit messages. Perhaps you should give an examples in an attempt to help people who are new to github or writing commit messages. I would vote for squashing because I believe most of the contributors did this during the contributions period and it might be hard to circle back with them in a reasonable time frame.

https://github.com/common-workflow-language/user_guide/blob/main/CONTRIBUTING.md#using-github links to https://docs.github.com/en/get-started/quickstart/github-flow which contains

> Give each commit a descriptive message to help you and future contributors understand what changes the commit contains. For example, fix typo or increase rate limit.
>
> Ideally, each commit contains an isolated, complete change. This makes it easy to revert your changes if you decide to take a different approach. For example, if you want to rename a variable and add some tests, put the variable rename in one commit and the tests in another commit. Later, if you want to keep the tests but revert the variable rename, you can revert the specific commit that contained the variable rename. If you put the variable rename and tests in the same commit or spread the variable rename across multiple commits, you would spend more effort reverting your changes.

Note that this norm is hardly unique to the CWL User Guide (including [the Arvados project](https://dev.arvados.org/projects/arvados/wiki/Coding_Standards#Git-commits)), so I would expect all mergers of PRs to either squash the PR down to one commit with a useful message (for smaller changes) or work with the contributor to reformat the commits and merge/rebase them.

[I don't know why your comment was left as a reply to an unrelated change; it makes the conversation hard to follow]

[the remark this answers is not in the document]
Comment (swzCuroverse) [reply src/faq.md]: I am not saying it is not a "norm" for development projects, I was just wanted to make sure that this was mentioned somewhere in the contributing guide.  If we think it is a big issue, it might be good to put it directly in the guide instead of hoping they will read the link and find it in the link.

**Comment by tetron** [comment]
@mr-c I believe branch protection on `main` is interfering with "Update branch" and "Commit suggestion" buttons, so I was going to close this pull request and make a new branch called `staging`.

**Comment by tetron** [comment]
See #373

File: src/faq.md [hunk 15]
```diff
@@ -79,7 +79,62 @@ outputs:
       outputEval: ${self[0].basename=inputs.otu_table_name; return self;}
 ```
 
-## Setting `self`-based input bindings for optional inputs
+## Referencing a Local Script
+
+There are two ways to reference a local script:
+
+The first method involves adding the folder containing your scripts to the `PATH` environment variable. This allows you to run the shell script without using `sh` or `bash` commands.
```
[the remark this answers is not in the document]
Comment (kinow) [reply src/faq.md]: Checked `main` and it contains the suggested text above

>The first method involves adding the folder containing your scripts to the PATH environment variable. This allows you to run the shell script directly without using sh or bash commands.

Marking as resolved.

File: src/faq.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (kinow) [reply src/faq.md]: I'll create an issue to add the contributing guide with instructions about squashing commits. Marking this thread as resolved.

File: src/faq.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (kinow) [reply src/faq.md:135]: Hmm, this comment seems a bit lost as it is right now. IMHO it could be re-written to better explain what it means, or removed. Will add a note to the staging PR.

File: src/faq.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (kinow) [reply src/faq.md:250]: I think the very first page of the user guide has "Command Line" in the table of contents. If you click next, there's no command* text, but the next page contains "command-line". So we are indeed not consistent at the moment. Follow-up issue: https://github.com/common-workflow-language/user_guide/issues/392

File: src/faq.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (kinow) [reply src/faq.md:322]: Hmm, `main` contains "How do I set mutually exclusive parameters?". So that's a question, but not 100% the suggested text. I'll create an issue for this just in case: https://github.com/common-workflow-language/user_guide/issues/393

File: src/introduction/prerequisites.md [hunk 16]
```diff
@@ -89,11 +88,14 @@ You can run the CWL tool description by omitting the `--validate` option:
 :caption: Running `true.cwl` with `cwltool`.
 ```
 
-### cwl-runner Python module
+### Cwl-runner Python Module
```
[the remark this answers is not in the document]
Comment (kinow) [reply src/introduction/prerequisites.md:91]: The current text may (very likely!) have been written by me :grimacing: The suggested text is less technical and easier to understand IMHO, so +1
